### PR TITLE
CLI to differentiate between 'Kafka source topic' and 'Kafka output topic' for CT/CS vs CTAS/CSAS

### PR DESCRIPTION
**Pull request by big-andy-coates**
Fix for #838 

Capture information in the metastore to allow CLI to know if this is a source topic behind a CT/CS statement, or a sink topic behind a CTAS/CSAS statement.

I did originally just switch the output to 'Kafka topic', i.e. dropped the 'output' part, as this was the easiest. However, @miguno requested we differentiate between the two.  This required exposing this information to the CLI.  I'm not overly happy with how I've exposed this. It feels like it should maybe be a field on KsqlTopic?  And maybe should be an enum for topic type e.g. {Source, Sink, Repartition, Changelog}.  Alternatively, the boolean could be change to just indicate this is a `CT/CS` statement, which would help with #837.

However, I thought I'd push this first version to get opinions before proceeding...

### CODE REVIEW

**Comment by hjafarpour** [comment]
After reading the discussion at #838 I think we need to clarify the difference between queries (computation) and data (Stream/Table/Topic). Similar to many query processing systems we use `EXPLAIN` to provide information about a query (computation) and `DESCRIBE` to provide information about data (metadata indeed). In such systems, it's query that depends on data and data itself is independent of query. So you can have data (Stream/Table/Topic) without having any query in the system. Therefore, it would not make sense to bring in the details on how the data is being used by a query/queries in description of data. 
The notion of `kafka input topic` and `kafka output topic` therefore would be misleading since a topic can be read by queries and also you can have queries that write into the same topic. I would just use `kafka topic` to represent the name of the kafka topic that backs the given stream or table.
We can keep track of the queries that read from and write into a stream or table and provide such information in `DESCRIBE EXTENDED`. This can be easily done after we merge the referential integrity PR (https://github.com/confluentinc/ksql/pull/652) since we will have such information in the metastore.

File: ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java [hunk 1]
```diff
@@ -477,26 +481,30 @@ private void printSchema(SourceDescription source) {
...
+                             : source.getTimestamp();
+
+    writer().println(String.format("%-20s : %s", "Key field", source.getKey()));
+    writer().println(String.format("%-20s : %s", "Key format", "STRING"));
+    writer().println(String.format("%-20s : %s", "Timestamp field", timestamp));
+    writer().println(String.format("%-20s : %s", "Value format", source.getSerdes()));
+
+    if (!source.getTopic().isEmpty()) {
+      final String title =
+          String.format("Kafka %s topic", source.isSinkTopic() ? "output" : "source");
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-cli/src/main/java/io/confluent/ksql/cli/console/Console.java]: Yer, I know what you mean. The more I've thought about this, the more I've come around to the same way of thinking.

**Comment by big-andy-coates** [comment]
Awesome, so we're looking at just using "Kafka topic"? @miguno are you happy with this?

(Much smaller change, so I'm very happy with that).

**Comment by miguno** [comment]
Let's go with "Kafka topic" for now.

**Comment by big-andy-coates** [comment]
OK, have backed out all the other work, (With the exception of some minor code clean up that I'd done along the way).  This is good for review again, please.

File: ksql-metastore/src/main/java/io/confluent/ksql/metastore/StructuredDataSource.java [hunk 2]
```diff
@@ -21,31 +21,30 @@
...
 import io.confluent.ksql.util.timestamp.TimestampExtractionPolicy;
 
 public abstract class StructuredDataSource implements DataSource {
 
-  final String dataSourceName;
-  final DataSourceType dataSourceType;
-  final Schema schema;
-  final Field keyField;
-  final TimestampExtractionPolicy timestampExtractionPolicy;
+  protected final String dataSourceName;
```
Comment (rodesai) [inline ksql-metastore/src/main/java/io/confluent/ksql/metastore/StructuredDataSource.java]: why not just leave these package private?

Comment (big-andy-coates) [reply ksql-metastore/src/main/java/io/confluent/ksql/metastore/StructuredDataSource.java]: protected gives better encapsulation.

**Pull request merged**